---
title: "Data Cleaning"
author: Monica N Mweetwa
date: 30/12/2024
---
Set working directory
```{r setup}
   knitr::opts_knit$set(root.dir = normalizePath("~/Dropbox/BEECH_16s_Analysis/Stunting_SAM")) 
   #knitr::opts_knit$set(root.dir = normalizePath("C:/Users/Monica/Dropbox/BEECH_16s_Analysis/Stunting_SAM")) 
```

1. Data Cleaning
```{r}
load("Data/RData/Zambia_EE_BEECH_16s_absQuant_forMonica.RData")
#this RData file contains:
#map - mapping file
#ps2 - phyloseq object with both BEECH and EEBiomarker data from WashU

library(readr)
library(data.table) ## to use %like% function
library(tidyverse)
#Import metadata files
SamBeechMetadata <- read_csv("Data/Metadata/SamBeechMetadata22.csv")
beech_morph <- read_csv("Data/Metadata/Morphometry/BEECHmorphometry_91_edited.csv")
sam_morph <- read_csv("Data/Metadata/Morphometry/SAM_morphometry.csv")

#Combine metadata for BEECH and EEBiomarker
beech_morph2 <- beech_morph %>% 
  mutate(HIV_status2 = ifelse(hiv == 2, 1, hiv)) %>%
  mutate_at(vars(c('VH', 'CD', 'ESA', 'HIV_status2')),
              as.numeric)
sam_morph2 <- sam_morph %>% rename('ESA' = 'vpmm') %>% mutate(PID = as.character(PID)) #Swap vars in rename function if using Windows computer
morph_dat <- full_join(sam_morph2, beech_morph2)
dat <- SamBeechMetadata %>%
    mutate_at(vars(c('AgeMonths_endo', 'WAZ_endo', 'LAZ_endo', 'WHZ_endo', 'HIV_status', 'GastricPh_endo', 'GLP.2_endo', 'I.FABP_endo', 'LPS_endo', 'LBP_endo', 'sCD14_endo')),
              as.numeric) %>%
  mutate(I.FABP_new = I.FABP_endo/1000,
         sCD14_new = sCD14_endo/1000) 
dat1 <- left_join(dat, morph_dat)
dat2 <- dat1 %>%
  mutate(HIV_status = ifelse(mal.type == "Stunting", HIV_status2, HIV_status)) %>%
  mutate(hiv_char = ifelse(HIV_status == 0, "Negative", "Positive"))
##rename the SAM children to match the data in the other files
dat2$SampleID <- ifelse(dat2$SampleID %like% "DryaspirateGC", dat2$SampleID, paste0("Zambia",dat2$SampleID)) 
write.csv(dat2, "Data/Metadata/Final_Metadata.csv")

#combine metadata with the mapping file from phyloseq object
map <- map %>%
  mutate(SampleID = sampleID)
map_table <- left_join(map, dat2)
map_tab <- map_table %>% 
  tibble::column_to_rownames("sampleID") 

#add new metadata table to phyloseq object
library(phyloseq)
sample_data(ps2) <- map_tab

#remove bacteria with  
ps0 = subset_taxa(ps2, Kingdom == "Bacteria") # remove those with no taxonomic assignment
ps0 = subset_taxa(ps0, Family != "Mitochondria") #remove those assigned as Mitochondria Family
ps0 = subset_taxa(ps0, Class != "Chloroplast") #remove those assigned as Chloroplast Class 
ps0

#Add sequence information to phyloseq object using the taxonomy tables
Taxtab<- ASVnames %>%
  rownames_to_column(var = "row_names") %>%
  select(row_names, ASVSeq) %>%
  mutate(seq.name = paste0(">",row_names)) %>%
  select(seq.name, ASVSeq) %>%
  rename('seq.text' = 'ASVSeq') %>% #Swap vars if using Windows computer
  filter(seq.name != ">ASV1_Alicyclobacillus acidiphilus") #Alicyclobacillus acidiphilus was a spike-in that should be excluded from analysis.
library(phylotools)
library(seqinr)
library(ape)
library(msa) 
library(stringr)
dat2fasta(Taxtab, outfile = "Data/Sequences.fasta")
seq <- readAAStringSet("Data/Sequences.fasta")   
## Perform multiple sequence alignment
my_alignment <- msa(seq)

## Compute distance matrix
my_alignment_sequence <- msaConvert(my_alignment, type="seqinr::alignment")

distance_alignment <- dist.alignment(my_alignment_sequence)

## compute phylogenetic tree using neighbor joining
Tree <- bionj(distance_alignment)
#Rename tip.labels of tree to match the names of phyloseq taxa
Tree2 <- Tree
Tree2$tip.label <- str_sub(Tree2$tip.label, start = 2)
write.tree(Tree2, "Data/Tree.tre")

phy_tree(ps0) <- Tree2
ps0

save(ps0, file = "Data/RData/phyloseq_dataset_species.RData")
```
